datashader/core.py

```diff
@@ -237,78 +237,42 @@ def trimesh(self, vertices, simplices, mesh=None, agg=None, interp=True):
-                                                'are required for the triangle '
-                                                'definition')
-        simplices_all_ints = simplices.dtypes.iloc[:3].map(
-            lambda dt: np.issubdtype(dt, np.integer)
-        ).all()
-        assert simplices_all_ints, ('Simplices must be integral. You may '
-                                    'consider casting simplices to integers '
-                                    'with ".astype(int)"')
-
-        # Choose reduction based on vertex coordinate data
-        weight_col = None
-        verts_have_weights = not not weights
-        if verts_have_weights:
-            weight_col = weights[0]
-            if agg is None:
-                agg = mean_rdn(weight_col)
-            elif agg.column is None:
-                # Rasterization code assumes presence of agg column
-                agg.column = weight_col
-        else:
-            assert simplices.values.shape[1] > 3, 'If no vertex weight column is provided, a triangle weight column is required.'
-            weight_col = simplices.columns[3]
-            if agg is None:
-                agg = mean_rdn(weight_col)
-            elif agg.column is None:
-                # Rasterization code assumes presence of agg column
-                agg.column = weight_col
+        source = mesh
 
-        if mesh is None:
+        # Validation is done inside the [pd]d_mesh utility functions
+        if source is None:
             if (isinstance(vertices, dd.DataFrame) and
                     isinstance(simplices, dd.DataFrame)):
-                # TODO: For dask: avoid .compute() calls, and add winding auto-detection
-                vertex_idxs = simplices.values[:, :3].astype(np.int64)
-                vals = vertices.values.compute()[vertex_idxs]
-                vals = vals.reshape(np.prod(vals.shape[:2]), vals.shape[2])
-                source = pd.DataFrame(vals, columns=vertices.columns)
-                if not verts_have_weights:
-                    source[weight_col] = simplices.values[:, 3].compute().repeat(3)
+                source = dd_mesh(vertices, simplices)
             else:
-                winding = [0, 1, 2]
-                # Change vertex winding to CW if necessary
-                first_tri = vertices.values[simplices.values[0, winding].astype(np.int64), :2]
-                a, b, c = first_tri
-                if np.cross(b-a, c-a).item() >= 0:
-                    winding = [0, 2, 1]
-
-                vertex_idxs = simplices.values[:, winding].astype(np.int64)
-                vals = vertices.values[vertex_idxs]
-                vals = vals.reshape(np.prod(vals.shape[:2]), vals.shape[2])
-                source = pd.DataFrame(vals, columns=vertices.columns)
-                if not verts_have_weights:
-                    source[weight_col] = simplices.values[:, 3].repeat(3)
+                source = pd_mesh(vertices, simplices)
+
```

Comment:
Can we move this code out as well, to be the user-level API for these functions, i.e.:

```
def mesh(vertices, simplices):
    """..."""
    if (isinstance(vertices, dd.DataFrame) and isinstance(simplices, dd.DataFrame)):
        return _dd_mesh(vertices, simplices)
    else:
        return _pd_mesh(vertices, simplices)
```

It seems like that name will be more obvious to the user, who normally doesn't have to think a lot about whether they are using pandas or dask dataframes.

Reply:
Sure, sounds reasonable. I'll have it done shortly.